src/Microsoft.NetCore.Analyzers/Core/Security/DoNotUseInsecureDeserializerWithoutBinderBase.cs

```diff
@@ -0,0 +1,165 @@
+
+            context.EnableConcurrentExecution();
+
+            // Security analyzer - analyze and report diagnostics on generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(
+                (CompilationStartAnalysisContext compilationStartAnalysisContext) =>
+                {
+                    WellKnownTypeProvider wellKnownTypeProvider =
+                    WellKnownTypeProvider.GetOrCreate(compilationStartAnalysisContext.Compilation);
+
+                    if (!wellKnownTypeProvider.TryGetKnownType(
+                            this.DeserializerTypeMetadataName,
+                            out INamedTypeSymbol deserializerTypeSymbol))
+                    {
+                        return;
+                    }
+
+                    compilationStartAnalysisContext.RegisterOperationBlockStartAction(
+                        (OperationBlockStartAnalysisContext operationBlockStartAnalysisContext) =>
+                        {
+                            ISymbol owningSymbol = operationBlockStartAnalysisContext.OwningSymbol;
+                            bool requiresBinderMustBeSetDataFlowAnalysis = false;
+
+                            operationBlockStartAnalysisContext.RegisterOperationAction(
+                                (OperationAnalysisContext operationAnalysisContext) =>
+                                {
+                                    IInvocationOperation invocationOperation =
+                                        (IInvocationOperation) operationAnalysisContext.Operation;
+                                    if (invocationOperation.TargetMethod.ContainingType == deserializerTypeSymbol
+                                        && cachedDeserializationMethodNames.Contains(invocationOperation.TargetMethod.Name))
+                                    {
+                                        requiresBinderMustBeSetDataFlowAnalysis = true;
+                                    }
+                                },
+                                OperationKind.Invocation);
+
+                            operationBlockStartAnalysisContext.RegisterOperationBlockEndAction(
+                                (OperationBlockAnalysisContext operationBlockAnalysisContext) =>
+                                {
+                                    if (!requiresBinderMustBeSetDataFlowAnalysis)
+                                    {
+                                        return;
+                                    }
+
+                                    ImmutableDictionary<IInvocationOperation, PropertySetAbstractValue> dfaResult =
+                                        PropertySetAnalysis.GetOrComputeHazardousParameterUsages(
+                                            operationBlockAnalysisContext.OperationBlocks[0].GetEnclosingControlFlowGraph(),
```

Comment:
>OperationBlocks[0] [](start = 74, length = 18)

Make sure you have unit tests for method with optional parameters, which will trigger multiple operation blocks for a method. I think the first one is always the method body block though, so this should be fine, but having a test is better :)

Reply:
Added.